README.md
# check_fritz

![CI](https://github.com/mcktr/check_fritz/workflows/CI/badge.svg?branch=master)

**Table of Contents**

1. Introduction
2. Support
3. Requirements
4. Installation
5. Usage

## Introduction

This is a check plugin, written in Go, for Icinga 2 to monitor a Fritz!Box.

## Support

Please ask your questions related to the configuration in the community channels. For bugs and feature requests 
please open an issue. 

## Requirements

The TR-064 feature must be enabled on the Fritz!Box.

You can enable the feature in the following menu:

```
Heimnetz -> Netzwerk -> Netzwerkeinstellungen ->  Heimnetzfreigaben -> Zugriff für Anwendungen zulassen
```

![Fritz!Box configuration](doc/images/fritzbox-configuration-tr064.png)

## Installation

* Download the latest stable binary package for your architecture from the [release page](https://github.com/mcktr/check_fritz/releases).
* Verify the checksum using `sha256sum`.

Example:

```
$ wget [DOWNLOAD URL CURRENT RELEASE]]/check_fritz.linux.amd64
$ wget [DOWNLOAD URL CURRENT RELEASE]]/check_fritz.linux.amd64.sha256

$ sha256sum -c check_fritz.linux.amd64.sha256
check_fritz.linux.amd64: OK
```

The checksum verification must return OK.

If you successfully verified the checksum, move the check plugin binary to your Icinga 2 plugin directory. It is also 
recommended to rename the binary package to `check_fritz`, since the suffix is only required for the checksum verification.

Ensure that you also set the executable bit with `chmod +x check_fritz`.

## Usage

### Parameter

| Parameter (short) | Parameter (long)     | Description                                                                                                       |
| ----------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------- |
| `-H`              | `--hostname`         | **Optional.** IP-Address or Hostname of the Fritz!Box. Defaults to `fritz.box`.                                   |
| `-P`              | `--port`             | **Optional.** Port for TR-064 over SSL. Defaults to `49443`.                                                      |
| `-u`              | `--username`         | **Optional.** Fritz!Box web interface Username for authentication. Defaults to `dslf-config`.                     |
| `-p`              | `--password`         | **Required.** Fritz!Box web interface Password for authentication.                                                |
| `-m`              | `--method`           | **Optional.** Defines the used check method. Defaults to `connection_status`.                                     |
| `-w`              | `--warning`          | **Optional.** Defines a warning threshold. Defaults to none.                                                      |
| `-c`              | `--critical`         | **Optional.** Defines a critical threshold. Defaults to none.                                                     |
| `-a`              | `--ain`              | **Optional.** Defines the AIN required by smart device check methods. Defaults to none.                           |
| `-t`              | `--timeout`          | **Optional.** Defines the timeout in seconds to wait for an answer from the Fritz!Box. Defaults to `90`.          |
| `-M`              | `--modelgroup`       | **Optional.** Defines the Fritz!Box model group. Supported model groups are `DSL` and `Cable`. Defaults to `DSL`. |
|                   | `--ignoresyncgroups` | **Optional.** Set a ignore list for SyncGroups (separate multiple values by comma (,).                            |
| `-d`              | `--debug`            | **Optional.** Enable debug output.                                                                                |

> **Note:**
>
> If you don't use the authentication method with username and password on your Fritz!Box, leave the username blank.


### Methods

| Name                     | Description                                                                 |
| ------------------------ | --------------------------------------------------------------------------- |
| `connection_status`      | WAN connection status.                                                      |
| `connection_uptime`      | WAN connection uptime in seconds.                                           |
| `device_uptime`          | Device uptime in seconds.                                                   |
| `device_update`          | Update state.                                                               |
| `downstream_max`         | Maximum downstream.                                                         |
| `downstream_current`     | Current downstream.                                                         |
| `downstream_usage`       | Current downstream usage.                                                   |
| `smart_heatertemperatur` | Current temperature of a radiator thermostat. Requires AIN.                 |
| `smart_socketpower`      | Current power consumption of a socket switch. Requires AIN.                 |
| `smart_socketenergy`     | Total power consumption of the last year of a socket switch.  Requires AIN. |
| `smart_status`           | Current smart device connection status. Requires AIN.                       |
| `upstream_max`           | Maximum upstream.                                                           |
| `upstream_current`       | Current upstream.                                                           |
| `upstream_usage`         | Current upstream usage.                                                     |

### Icinga 2 CheckCommand

You can use this predefined Icinga 2 CheckCommand. For more information about CheckCommand objects please refere to 
the official [Icinga 2 documenation](https://icinga.com/docs/icinga2/latest/). 

```
object CheckCommand "fritz" {
	command = [ CustomPluginDir + "/check_fritz" ]

	arguments += {
		"--hostname" = "$fritz_hostname$"
		"--port" = "$fritz_port$"
		"--username" = "$fritz_username$"
		"--password" = "$fritz_password$"
		"--method" = "$fritz_method$"
		"--warning" = "$fritz_warning$"
		"--critical" = "$fritz_critical$"
		"--index" = "$fritz_index$"
		"--ain" = "$fritz_ain$"
		"--timeout" = "$fritz_timeout$"
		"--modelgroup" = "$fritz_modelgroup$" 
	}

	vars.fritz_hostname = "$address$"
}
```

> **Note**
>
> This predefined CheckCommand object for Icinga 2 requires a global constants named `CustomPluginDir`.
> You can define them in `constants.conf` in the root Icinga 2 configuration directory. For more information
> please read the official Icinga 2 documentation about [global constants](https://icinga.com/docs/icinga2/latest/doc/04-configuring-icinga-2/#constants-conf).

### Thanks

Thanks to all contributors!

* [fdellwing](https://github.com/fdellwing) for [suggesting the upstream_/downstream_usage functions](https://github.com/mcktr/check_fritz/issues/48)
* [uclara](https://github.com/uclara) for [testing the support for Fritz!Box cable models](https://github.com/mcktr/check_fritz/issues/40)


CHANGELOG.md
# Changelog

## 1.2.2 (2020-11-14)

- Bugfix release

* [#106](https://github.com/mcktr/check_fritz/pull/106) (core/build, bug): Fix checksum creation during the build process

## 1.2.1 (2020-11-07)

- Bugfix release

* [#104](https://github.com/mcktr/check_fritz/pull/104) (area/check/connection, area/performance-data): Fix connection_status performance data output 
* [#103](https://github.com/mcktr/check_fritz/pull/103) (area/check/connection, area/performance-data, bug): Missing performance data output in connection_status & device_update methods

## 1.2.0 (2020-10-09)

- Please have a look into the [upgrading documentation](https://github.com/mcktr/check_fritz/blob/master/doc/upgrading.md).

* [#99](https://github.com/mcktr/check_fritz/pull/99) (area/cli, bug): Add missing smart_socketenergy entry to help page
* [#98](https://github.com/mcktr/check_fritz/pull/98) (area/cli, bug): Fix --modelgroup parameter alias not working
* [#97](https://github.com/mcktr/check_fritz/pull/97) (bug, area/check/downstream, area/check/upstream): Fix wrong values for max down-/upstream
* [#95](https://github.com/mcktr/check_fritz/pull/95) (area/documentation): Update contributing documentation
* [#92](https://github.com/mcktr/check_fritz/pull/92) (area/documentation): Add upgrading documentation for v1.2.0
* [#91](https://github.com/mcktr/check_fritz/pull/91) (area/check/smart-device, bug): Fix nil pointer dereference in smart check methods
* [#90](https://github.com/mcktr/check_fritz/pull/90) (area/cli, bug): Fix typo in help output
* [#87](https://github.com/mcktr/check_fritz/pull/87) (core/quality, enhancement): Enhance the build process
* [#85](https://github.com/mcktr/check_fritz/pull/85) (core/quality, area/documentation): Add note for SELinux context
* [#81](https://github.com/mcktr/check_fritz/pull/81) (core/quality, enhancement): Add CI artifacts upload
* [#80](https://github.com/mcktr/check_fritz/pull/80) (core/quality, enhancement): Add debug mode parameter
* [#76](https://github.com/mcktr/check_fritz/pull/76) (core/quality, enhancement): Replace cmdline package with cli (v2) package
* [#70](https://github.com/mcktr/check_fritz/pull/70) (core/quality): Make internal exit code variables to constants
* [#69](https://github.com/mcktr/check_fritz/pull/69) (core/quality): Remove deprecated smart device functions

## 1.1.0 (2019-09-18)

- Please have a look into the [upgrading documentation](https://github.com/mcktr/check_fritz/blob/master/doc/upgrading.md).

* [#65](https://github.com/mcktr/check_fritz/pull/65) (area/check/downstream, area/check/upstream, bug): Fix dividing by zero
* [#64](https://github.com/mcktr/check_fritz/pull/64) (area/documentation): Introduce upgrading documentation
* [#63](https://github.com/mcktr/check_fritz/pull/63) (area/documentation, core/quality): Add development documentation
* [#62](https://github.com/mcktr/check_fritz/pull/62) (area/check/connection, enhancement): Implement support for Fritz!Box cable models
* [#58](https://github.com/mcktr/check_fritz/pull/58) (area/cli, core/quality): Implement timeout parameter
* [#60](https://github.com/mcktr/check_fritz/pull/60) (area/documentation, bug): Fix spelling in README
* [#59](https://github.com/mcktr/check_fritz/pull/59) (area/cli, bug): Remove default values in help text output
* [#57](https://github.com/mcktr/check_fritz/pull/57) (area/soap, bug, core/build-fix): Fix panic when Fritz!Box is unreachable
* [#55](https://github.com/mcktr/check_fritz/pull/55) (area/check/downstream, area/check/upstream, enhancement): Implement method for up-/downstream utilization
* [#54](https://github.com/mcktr/check_fritz/pull/54) (area/documentation): Adjust documentation for next release
* [#51](https://github.com/mcktr/check_fritz/pull/51) (area/check/smart-device, bug): Fix smart device index handling
* [#50](https://github.com/mcktr/check_fritz/pull/50) (core/quality, enhancement): Rewrite check plugin to use Goroutines
* [#47](https://github.com/mcktr/check_fritz/pull/47) (core/quality): Use Go modules to handle dependencies
* [#45](https://github.com/mcktr/check_fritz/pull/45) (enhancement): Implement check method for smart device status
* [#44](https://github.com/mcktr/check_fritz/pull/44) (enhancement): Add external IP to status method
* [#39](https://github.com/mcktr/check_fritz/pull/39) (bug): Fix CHANGELOG format

## 1.0 (2019-03-22)

- This is the first stable version of check_fritz

* [#38](https://github.com/mcktr/check_fritz/pull/38) (documentation): Add installation documentation
* [#37](https://github.com/mcktr/check_fritz/pull/37) (enhancement): Rename check methods to their final names
* [#36](https://github.com/mcktr/check_fritz/pull/36) (enhancement): Add --version command-line option
* [#35](https://github.com/mcktr/check_fritz/pull/35) (build): Provide SHA-256 checksum
* [#34](https://github.com/mcktr/check_fritz/pull/34) (build, tests): Add Go 1.12 to the CI tests
* [#33](https://github.com/mcktr/check_fritz/pull/33) (tests): Fix golint import path in CI tests
* [#31](https://github.com/mcktr/check_fritz/pull/31) (bug): Make password optional
* [#30](https://github.com/mcktr/check_fritz/pull/30) (bug): Fix wrong performance data value in socketpower
* [#29](https://github.com/mcktr/check_fritz/pull/29) (enhancement): Calculate human readable uptime
* [#28](https://github.com/mcktr/check_fritz/pull/28) (bug): Fix wrong new line statement in output
* [#27](https://github.com/mcktr/check_fritz/pull/27) (bug): Fix performance data output in downstream_max
* [#26](https://github.com/mcktr/check_fritz/pull/26) (bug): Fix null pointer reference
* [#24](https://github.com/mcktr/check_fritz/pull/24) (bug): Make Golint happy
* [#23](https://github.com/mcktr/check_fritz/pull/23) (enhancement): Add Travis CI integration
* [#22](https://github.com/mcktr/check_fritz/pull/22) (code-quality): Use go-cmdline library
* [#21](https://github.com/mcktr/check_fritz/pull/21) (code-quality): Make Golint happier

cmd/check_fritz/main.go
package main

import (
	"fmt"
	"os"

	"github.com/urfave/cli/v2"
)

// program version
var Version = "1.2.2"

// internal exit codes
const (
	exitOk       = 0
	exitWarning  = 1
	exitCritical = 2
	exitUnknown  = 3
)

// GlobalReturnCode holds always the last set return code
var GlobalReturnCode = exitUnknown

// ArgumentInformation is the data structure for the passed arguments
type ArgumentInformation struct {
	Hostname      *string
	Port          *string
	Username      *string
	Password      *string
	Method        *string
	Warning       *float64
	Critical      *float64
	InputVariable *string
	Timeout       *int
	Modelgroup    *string
	Debug         bool
}

func createRequiredArgumentInformation(hostname string, port string, username string, password string, method string, timeout int, modelgroup string) ArgumentInformation {
	var ai ArgumentInformation

	ai.Hostname = &hostname
	ai.Port = &port
	ai.Username = &username
	ai.Password = &password
	ai.Method = &method
	ai.Modelgroup = &modelgroup
	ai.Timeout = &timeout
	ai.Debug = false

	return ai
}

func (ai *ArgumentInformation) createWarningThreshold(warning float64) {
	ai.Warning = &warning
}

func (ai *ArgumentInformation) createCriticalThreshold(critical float64) {
	ai.Critical = &critical
}

func (ai *ArgumentInformation) createInputVariable(v string) {
	ai.InputVariable = &v
}

func (ai *ArgumentInformation) setDebugMode() {
	ai.Debug = true
}

func printVersion() {
	fmt.Println("check_fritz v" + Version)
	GlobalReturnCode = exitOk
}

func checkRequiredFlags(aI *ArgumentInformation) bool {
	if aI.Hostname == nil || *aI.Hostname == "" {
		fmt.Println("No hostname")
		return false
	}

	if aI.Port == nil || *aI.Port == "" {
		fmt.Println("No port")
		return false
	}

	if aI.Username == nil || *aI.Username == "" {
		fmt.Println("No username")
		return false
	}

	return true
}

// HandleError is the global error handler for the programm
func HandleError(err error) bool {
	if err != nil {
		fmt.Println(err)
		GlobalReturnCode = exitUnknown
		return true
	}

	return false
}

func checkMain(c *cli.Context) error {

	type param struct {
		Hostname   string
		Port       string
		Username   string
		Password   string
		Method     string
		Timeout    int
		ModelGroup string
	}

	p := param{}

	p.Hostname = c.String("hostname")
	p.Port = c.String("port")
	p.Username = c.String("username")

	if c.IsSet("password") {
		p.Password = c.String("password")
	}

	p.Method = c.String("method")
	p.Timeout = c.Int("timeout")
	p.ModelGroup = c.String("modelgroup")

	argInfo := createRequiredArgumentInformation(p.Hostname, p.Port, p.Username, p.Password, p.Method, p.Timeout, p.ModelGroup)

	if c.IsSet("warning") {
		argInfo.createWarningThreshold(c.Float64("warning"))
	}

	if c.IsSet("critical") {
		argInfo.createCriticalThreshold(c.Float64("critical"))
	}

	if c.IsSet("ain") {
		argInfo.createInputVariable(c.String("ain"))
	}

	if c.IsSet("debug") {
		argInfo.setDebugMode()
	}

	if !checkRequiredFlags(&argInfo) {
		os.Exit(exitUnknown)
	}

	switch *argInfo.Method {
	case "connection_status":
		CheckConnectionStatus(argInfo)
	case "connection_uptime":
		CheckConnectionUptime(argInfo)
	case "device_uptime":
		CheckDeviceUptime(argInfo)
	case "device_update":
		CheckDeviceUpdate(argInfo)
	case "downstream_max":
		CheckDownstreamMax(argInfo)
	case "upstream_max":
		CheckUpstreamMax(argInfo)
	case "downstream_current":
		CheckDownstreamCurrent(argInfo)
	case "downstream_usage":
		CheckDownstreamUsage(argInfo)
	case "upstream_usage":
		CheckUpstreamUsage(argInfo)
	case "upstream_current":
		CheckUpstreamCurrent(argInfo)
	case "smart_heatertemperatur":
		CheckSpecificSmartHeaterTemperatur(argInfo)
	case "smart_socketpower":
		CheckSpecificSmartSocketPower(argInfo)
	case "smart_socketenergy":
		CheckSpecificSmartSocketEnergy(argInfo)
	case "smart_status":
		CheckSpecificSmartStatus(argInfo)
	default:
		fmt.Println("Unknown method.")
		GlobalReturnCode = exitUnknown
	}

	os.Exit(GlobalReturnCode)

	return nil
}

func main() {
	app := &cli.App{
		Action:  checkMain,
		Name:    "check_fritz",
		Usage:   "Check plugin to monitor a Fritz!Box",
		Version: Version,
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:        "hostname",
				Aliases:     []string{"H"},
				Value:       "fritz.box",
				DefaultText: "fritz.box",
				Usage:       "Specifies the hostname.",
			},
			&cli.StringFlag{
				Name:        "port",
				Aliases:     []string{"P"},
				Value:       "49443",
				DefaultText: "49443",
				Usage:       "Specifies the SSL port.",
			},
			&cli.StringFlag{
				Name:        "username",
				Aliases:     []string{"u"},
				Value:       "dslf-config",
				DefaultText: "dslf-config",
				Usage:       "Specifies the username.",
			},
			&cli.StringFlag{
				Name:    "password",
				Aliases: []string{"p"},
				Usage:   "Specifies the password.",
			},
			&cli.StringFlag{
				Name:        "method",
				Aliases:     []string{"m"},
				Value:       "connection_status",
				DefaultText: "connection_status",
				Usage:       "Specifies the check method.",
			},
			&cli.StringFlag{
				Name:    "ain",
				Aliases: []string{"a"},
				Usage:   "Specifies the AIN for smart devices.",
			},
			&cli.IntFlag{
				Name:        "timeout",
				Aliases:     []string{"t"},
				Value:       90,
				DefaultText: "90",
				Usage:       "Specifies the timeout for requests.",
			},
			&cli.StringFlag{
				Name:        "modelgroup",
				Value:       "DSL",
				DefaultText: "DSL",
				Aliases:     []string{"M"},
				Usage:       "Specifies the Fritz!Bpx model group (DSL or Cable).",
			},
			&cli.Float64Flag{
				Name:    "warning",
				Aliases: []string{"w"},
				Usage:   "Specifies the warning threshold.",
			},
			&cli.Float64Flag{
				Name:    "critical",
				Aliases: []string{"c"},
				Usage:   "Specifies the critical threshold.",
			},
			&cli.BoolFlag{
				Name:    "debug",
				Aliases: []string{"d"},
				Usage:   "Outputs debug information",
			},
		},
	}

	cli.AppHelpTemplate = `NAME:

	{{.Name}} - {{.Usage}}

USAGE:
   check_fritz [options...]

OPTIONS:

	{{range .VisibleFlags}}{{.}}
	{{end}}

METHODS:
	connection_status       WAN connection status,
	connection_uptime       WAN connection uptime (in seconds),
	device_uptime           device uptime (in seconds),
	device_update           update state,
	downstream_max          maximum downstream,
	upstream_max            maximum downstream,
	downstream_current      current downstream,
	upstream_current        current upstream,
	downstream_usage        current downstream usage,
	upstream_usage          current upstream usage,
	smart_heatertemperatur  current temperature of a a radiator thermostat (requires AIN),
	smart_socketpower       current power consumption of a socket switch (requires AIN),
	smart_socketenergy 		Total power consumption of the last year of a socket switch (requires AIN),	
	smart_status            current smart device status (requires AIN)
`

	cli.VersionFlag = &cli.BoolFlag{
		Name: "version", Aliases: []string{"V"},
		Usage: "print the version",
	}

	app.Run(os.Args)
}
